### PR TITLE
Cut v1.2.0

### PR DESCRIPTION
**Pull request by beorn7**
It's about time.

@brian-brazil @fabxc @juliusv @grobie all good? Anything else that has to go in?

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,43 @@
+## 1.2.0 / 2016-10-07
+
+* [FEATURE] Enable the use of query parameters in `/graph` URLs.
+* [FEATURE] PromQL: Add `minute()` function.
+* [FEATURE] Add GCE service discovery.
+* [FEATURE] Allow job names to start with a digit.
```
Comment (juliusv) [inline CHANGELOG.md]: IIRC we actually changed that PR to allow arbitrary job names now.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,43 @@
+## 1.2.0 / 2016-10-07
+
+* [FEATURE] Enable the use of query parameters in `/graph` URLs.
```
Comment (juliusv) [inline CHANGELOG.md]: Maybe more like "Cleaner encoding of query parameters in `/graph` URLs" or so?

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,43 @@
...
+* [FEATURE] PromQL: Add `minute()` function.
+* [FEATURE] Add GCE service discovery.
+* [FEATURE] Allow job names to start with a digit.
+* [FEATURE] Allow disabling local storage.
+* [FEATURE] EC2 service discovery: Expose `ec2_instance_state`.
+* [FEATURE] HTTP basic auth and TLS support for remote write path.
+* [FEATURE] Support for relabelling in remote write path.
```
Comment (juliusv) [inline CHANGELOG.md]: Just a heads-up: this and remote write configuration in general aren't documented yet. Remote write in general should be documented as experimental. I filed two issues:

https://github.com/prometheus/docs/issues/582
https://github.com/prometheus/docs/issues/583

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,43 @@
...
+* [CLEANUP] Storage: Chunk code in own package. Added Utilization method.
+* [CLEANUP] Export various identifiers for better code reuse.
+* [CLEANUP] Cleanup code issues reported by goreport and added link to
+  goreport in README.md.
+* [CLEANUP] Various other minor code cleanups and fixed typos.
+
+Below changes are not considered breaking because they affect parts marked as
+experimental or unstable:
+
+* [CHANGE] Remote write path does not use gRPC anymore.
```
Comment (juliusv) [inline CHANGELOG.md]: Also, the remote write path is now configured via the configuration file, not via a flag anymore.

File: CHANGELOG.md (hunk 1 above)
Comment (grobie) [inline CHANGELOG.md]: Yes. I'd write "Allow all valid label values as job names in config files."

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,43 @@
...
+
+* [FEATURE] Enable the use of query parameters in `/graph` URLs.
+* [FEATURE] PromQL: Add `minute()` function.
+* [FEATURE] Add GCE service discovery.
+* [FEATURE] Allow job names to start with a digit.
+* [FEATURE] Allow disabling local storage.
+* [FEATURE] EC2 service discovery: Expose `ec2_instance_state`.
+* [FEATURE] HTTP basic auth and TLS support for remote write path.
+* [FEATURE] Support for relabelling in remote write path.
+* [ENHANCEMENT] Add `go_import_path` to Travis so it works on a fork.
```
Comment (grobie) [inline CHANGELOG.md]: This is a build internal and I wouldn't mention it here, as it doesn't help me as operator when reviewing a version update changelog.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,43 @@
...
+* [ENHANCEMENT] Makefile: Promu installation now logged.
```
Comment (grobie) [inline CHANGELOG.md]: Dito.

File: CHANGELOG.md (hunk 5 above)
Comment (grobie) [inline CHANGELOG.md]: If you believe it's important to mention such commits, I'd use a different group for such changes and move it to the bottom.

File: CHANGELOG.md (hunk 4 above)
Comment (fabxc) [inline CHANGELOG.md]: Still under "experimental" scope, right?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,43 @@
...
+* [BUGFIX] Handle NaN in `changes()` correctly.
+* [CLEANUP] Remove DNS resolution dependency from tests.
+* [CLEANUP] More tests for PromQL time functions.
+* [CLEANUP] Add contexts to PromQL and storage interfaces.
+* [CLEANUP] Consolidate web API contexts. Update `common/route` vendoring.
+* [CLEANUP] Storage: Chunk code in own package. Added Utilization method.
+* [CLEANUP] Export various identifiers for better code reuse.
+* [CLEANUP] Cleanup code issues reported by goreport and added link to
+  goreport in README.md.
+* [CLEANUP] Various other minor code cleanups and fixed typos.
```
Comment (fabxc) [inline CHANGELOG.md]: Irrelevant to the user.

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,43 @@
...
+* [BUGFIX] Do not fail Consul discovery on Prometheus startup when Consul
+  is down.
+* [BUGFIX] Handle NaN in `changes()` correctly.
+* [CLEANUP] Remove DNS resolution dependency from tests.
+* [CLEANUP] More tests for PromQL time functions.
+* [CLEANUP] Add contexts to PromQL and storage interfaces.
+* [CLEANUP] Consolidate web API contexts. Update `common/route` vendoring.
+* [CLEANUP] Storage: Chunk code in own package. Added Utilization method.
+* [CLEANUP] Export various identifiers for better code reuse.
+* [CLEANUP] Cleanup code issues reported by goreport and added link to
```
Comment (fabxc) [inline CHANGELOG.md]: Irrelevant to the user.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,43 @@
...
+* [BUGFIX] Fix race between reload and stop of the rule manager.
+* [BUGFIX] Do not fail Consul discovery on Prometheus startup when Consul
+  is down.
+* [BUGFIX] Handle NaN in `changes()` correctly.
+* [CLEANUP] Remove DNS resolution dependency from tests.
+* [CLEANUP] More tests for PromQL time functions.
+* [CLEANUP] Add contexts to PromQL and storage interfaces.
+* [CLEANUP] Consolidate web API contexts. Update `common/route` vendoring.
+* [CLEANUP] Storage: Chunk code in own package. Added Utilization method.
+* [CLEANUP] Export various identifiers for better code reuse.
```
Comment (fabxc) [inline CHANGELOG.md]: Irrelevant to the user. Only package users might be interested, who are not tied to the versioning scheme of the binary releases.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,43 @@
...
+* [BUGFIX] Update govalidator vendoring to fix URL validation.
+* [BUGFIX] Fix race between reload and stop of the rule manager.
+* [BUGFIX] Do not fail Consul discovery on Prometheus startup when Consul
+  is down.
+* [BUGFIX] Handle NaN in `changes()` correctly.
+* [CLEANUP] Remove DNS resolution dependency from tests.
+* [CLEANUP] More tests for PromQL time functions.
+* [CLEANUP] Add contexts to PromQL and storage interfaces.
+* [CLEANUP] Consolidate web API contexts. Update `common/route` vendoring.
+* [CLEANUP] Storage: Chunk code in own package. Added Utilization method.
```
Comment (fabxc) [inline CHANGELOG.md]: Irrelevant to the user. Only package users might be interested, who are not tied to the versioning scheme of the binary releases.

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,43 @@
...
+* [BUGFIX] Remove JSON from HTTP Accept header.
+* [BUGFIX] Update govalidator vendoring to fix URL validation.
+* [BUGFIX] Fix race between reload and stop of the rule manager.
+* [BUGFIX] Do not fail Consul discovery on Prometheus startup when Consul
+  is down.
+* [BUGFIX] Handle NaN in `changes()` correctly.
+* [CLEANUP] Remove DNS resolution dependency from tests.
+* [CLEANUP] More tests for PromQL time functions.
+* [CLEANUP] Add contexts to PromQL and storage interfaces.
+* [CLEANUP] Consolidate web API contexts. Update `common/route` vendoring.
```
Comment (fabxc) [inline CHANGELOG.md]: Irrelevant to the user.

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,3 +1,43 @@
...
+* [BUGFIX] Zookeeper service discovery: Resync state after Zookeeper failure.
+* [BUGFIX] Remove JSON from HTTP Accept header.
+* [BUGFIX] Update govalidator vendoring to fix URL validation.
+* [BUGFIX] Fix race between reload and stop of the rule manager.
+* [BUGFIX] Do not fail Consul discovery on Prometheus startup when Consul
+  is down.
+* [BUGFIX] Handle NaN in `changes()` correctly.
+* [CLEANUP] Remove DNS resolution dependency from tests.
+* [CLEANUP] More tests for PromQL time functions.
+* [CLEANUP] Add contexts to PromQL and storage interfaces.
```
Comment (fabxc) [inline CHANGELOG.md]: Irrelevant to the user.

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,3 +1,43 @@
...
+* [BUGFIX] Zookeeper service discovery: Remove deleted nodes.
+* [BUGFIX] Zookeeper service discovery: Resync state after Zookeeper failure.
+* [BUGFIX] Remove JSON from HTTP Accept header.
+* [BUGFIX] Update govalidator vendoring to fix URL validation.
+* [BUGFIX] Fix race between reload and stop of the rule manager.
+* [BUGFIX] Do not fail Consul discovery on Prometheus startup when Consul
+  is down.
+* [BUGFIX] Handle NaN in `changes()` correctly.
+* [CLEANUP] Remove DNS resolution dependency from tests.
+* [CLEANUP] More tests for PromQL time functions.
```
Comment (fabxc) [inline CHANGELOG.md]: Irrelevant to the user.

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,3 +1,43 @@
...
+* [BUGFIX] Fix word alignment of struct fields in tests.
+* [BUGFIX] Zookeeper service discovery: Remove deleted nodes.
+* [BUGFIX] Zookeeper service discovery: Resync state after Zookeeper failure.
+* [BUGFIX] Remove JSON from HTTP Accept header.
+* [BUGFIX] Update govalidator vendoring to fix URL validation.
+* [BUGFIX] Fix race between reload and stop of the rule manager.
+* [BUGFIX] Do not fail Consul discovery on Prometheus startup when Consul
+  is down.
+* [BUGFIX] Handle NaN in `changes()` correctly.
+* [CLEANUP] Remove DNS resolution dependency from tests.
```
Comment (fabxc) [inline CHANGELOG.md]: Irrelevant to the user.

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,3 +1,43 @@
...
+* [FEATURE] PromQL: Add `minute()` function.
+* [FEATURE] Add GCE service discovery.
+* [FEATURE] Allow job names to start with a digit.
+* [FEATURE] Allow disabling local storage.
+* [FEATURE] EC2 service discovery: Expose `ec2_instance_state`.
+* [FEATURE] HTTP basic auth and TLS support for remote write path.
+* [FEATURE] Support for relabelling in remote write path.
+* [ENHANCEMENT] Add `go_import_path` to Travis so it works on a fork.
+* [ENHANCEMENT] Makefile: Promu installation now logged.
+* [ENHANCEMENT] Fingerprint locking: Avoid having contended mutexes on
```
Comment (fabxc) [inline CHANGELOG.md]: Very implementation specific. Not every user will understand what this means for them in practice.

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,3 +1,43 @@
...
+  same cacheline.
+* [ENHANCEMENT] Avoid slow `defer` in the `seriesMap.get` hot spot.
```
Comment (fabxc) [inline CHANGELOG.md]: Very implementation specific. Not every user will understand what this means for them in practice.
Possibly consolidate the performance improvements into a single understandable entry.

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,3 +1,43 @@
...
+* [ENHANCEMENT] Improve label matching performance.
+* [BUGFIX] Fix word alignment of struct fields in tests.
+* [BUGFIX] Zookeeper service discovery: Remove deleted nodes.
+* [BUGFIX] Zookeeper service discovery: Resync state after Zookeeper failure.
+* [BUGFIX] Remove JSON from HTTP Accept header.
+* [BUGFIX] Update govalidator vendoring to fix URL validation.
```
Comment (fabxc) [inline CHANGELOG.md]: For the user this means fixed Alertmanager URL flags, which is not clear.

**Comment by beorn7** [comment]
> Sorry, this got quite spammy but I recall us having agreed long ago to only list changes that affect operators and users long ago.

I do not recall anything like that. I (strongly) think the changelog should document all changes, for all audiences (operators, builders, developers, users, …). Obviously, I wouldn't document "added a period at the end of doc comment", in those cases we should make the call and just summarize it as "various code cleanups", as I have done here.

As usual, I'm willing to bend to a majority decision of qualified people. Before changing anything in this PR, could we have a quick straw poll what the target audience of the changelog should be? @fabxc and myself have already cast their vote. @grobie seems to be on the same train as @fabxc (which would hint towards the existence of such an undocumented agreement). (Although I have to say that the comment @grobie was referring to is actually relevant to operators, should said operators maintain a fork of Prometheus, which might be common in larger organizations that audit code and build their own binaries from it.)

**Comment by brian-brazil** [comment]
Changelog should be aimed primarily at users. If someone wants a blog by blow they can look at the commit history.

My blog being the only usable source of user-relevant changelog info would not be good for the project IMHO.

**Comment by fabxc** [comment]
I'm confident anyone maintaining an internal fork and audits changes will not trust a changelog regardless of its verbosity. They also know how to run `git log --oneline --no-merges`, which is the only relevant change log for developers.

Yes, it does include a few commits adding periods to doc comments or whatnot – those are a pollution of history and waste of everyone's time in the first place. (Dozens of people are emailed about it, and at least 5 of them are taking action on each of them.)

**Comment by grobie** [comment]
As someone who has to regularly review and apply updates of software, I
feel strongly about only documenting user relevant changes like behaviour+
performance changes, features and bug fixes.

Changes to internal code structure, test setup, build infrastructure, etc.
are only relevant to the very small subset of active regular contributors
of which we have maybe 10 people. I disagree that the line I commented
about is relevant to the average operator.  So far the only reported use
case was for the single active public fork we have.

The small active developer audience continuously being interested in such
other changes and not already following the development can use git diff.

**Comment by beorn7** [comment]
Thanks for casting your votes. I think we have quorum. I'll update the documentation and this PR accordingly.

File: CHANGELOG.md (hunk 1 above)
Comment (beorn7) [inline CHANGELOG.md]: Done.

File: CHANGELOG.md (hunk 2 above)
Comment (beorn7) [inline CHANGELOG.md]: Done.

File: CHANGELOG.md (hunk 3 above)
Comment (beorn7) [inline CHANGELOG.md]: I marked those features as belonging to an experimental part of the project.

File: CHANGELOG.md (hunk 4 above)
Comment (beorn7) [inline CHANGELOG.md]: Added.

File: CHANGELOG.md (hunk 7 above)
Comment (beorn7) [inline CHANGELOG.md]: Removed as per general discussion.

File: CHANGELOG.md (hunk 15 above)
Comment (beorn7) [inline CHANGELOG.md]: Removed as per general discussion.

File: CHANGELOG.md (hunk 16 above)
Comment (beorn7) [inline CHANGELOG.md]: Consolidated into "Various performance improvements in local storage."

File: CHANGELOG.md (hunk 17 above)
Comment (beorn7) [inline CHANGELOG.md]: Changed.

File: CHANGELOG.md (hunk 14 above)
Comment (beorn7) [inline CHANGELOG.md]: Removed as per general discussion.

File: CHANGELOG.md (hunk 13 above)
Comment (beorn7) [inline CHANGELOG.md]: Removed as per general discussion.

File: CHANGELOG.md (hunk 12 above)
Comment (beorn7) [inline CHANGELOG.md]: Removed as per general discussion.

File: CHANGELOG.md (hunk 11 above)
Comment (beorn7) [inline CHANGELOG.md]: Removed as per general discussion.

File: CHANGELOG.md (hunk 10 above)
Comment (beorn7) [inline CHANGELOG.md]: Removed as per general discussion.

File: CHANGELOG.md (hunk 9 above)
Comment (beorn7) [inline CHANGELOG.md]: Removed as per general discussion.

File: CHANGELOG.md (hunk 8 above)
Comment (beorn7) [inline CHANGELOG.md]: Removed as per general discussion.

**Comment by beorn7** [comment]
CHANGELOG abbreviated according to new guideline and all other comments addressed.

I also removed again the separate CHANGES section for experimental features. The intention was to not have a big and fat breaking change on top of the list (for a minor release). Now I have grouped all changes to the remote write path at the bottom of the list and marked them cleary as experimental so people don't get scared about the breaking changes.

**Comment by beorn7** [comment]
Thanks everybody. Now working on the remaining steps for the release.

**Pull request merged**